// 0xorderbuilder/lib/services/SignService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const _0x_js_1 = require("0x.js");
const etherUtil = require("ethereumjs-util");
class SignService {
    static signOrder(order, privateKey) {
        const orderHash = _0x_js_1.ZeroEx.getOrderHashHex(order);
        const signature = etherUtil.ecsign(etherUtil.hashPersonalMessage(etherUtil.toBuffer(orderHash)), etherUtil.toBuffer(privateKey));
        return {
            ecSignature: {
                r: etherUtil.bufferToHex(signature.r),
                s: etherUtil.bufferToHex(signature.s),
                v: signature.v,
            },
            exchangeContractAddress: order.exchangeContractAddress,
            expirationUnixTimestampSec: order.expirationUnixTimestampSec,
            feeRecipient: order.feeRecipient,
            maker: order.maker,
            makerFee: order.makerFee,
            makerTokenAddress: order.makerTokenAddress,
            makerTokenAmount: order.makerTokenAmount,
            salt: order.salt,
            taker: order.taker,
            takerFee: order.takerFee,
            takerTokenAddress: order.takerTokenAddress,
            takerTokenAmount: order.takerTokenAmount,
        };
    }
}
exports.SignService = SignService;
